Add `CreatePullRequest` action

tommyming opened the pull request:
Closes #42217 

Release Notes:
- Added `git: create pull request` command to open a pull request for the current branch

For more details, please refer to the issue, thank you.

tommyming committed:
update create PR implementation with command and button support.

tommyming committed:
update button layout style to sync with fetch action, add disable state
when there are no PRs available.

tommyming committed:
update display text

tommyming committed:
remove create pull request button, leave only command option, clean
clippy errors and earnings.

tommyming committed:
update formatting

tommyming committed:
code cleanup, remove excessive implementation of the button.

tommyming commented:
@esthertrapadoux Thanks for the notice!
Didn't expect that this PR got into Quality Week!
Please take your time and keep up the good work!

dinocosta reviewed (changes requested):
Hey @tommyming  , thank you for opening this. This is a helpful feature, and the test inclusion is appreciated! ✌️

Overall the changes look good, but there are a couple of places where we can improve the implementation and a small edge case we’re not handling correctly. Let me know if you have any questions.

tommyming committed:
remove `BuildCreatePullRequestParams`,
update pull request creation handling: check upstream branch first, local branch as fallback,
code cleanup, follow coding styles.

dinocosta commented on crates/git_ui/src/git_panel.rs:
```diff
@@ -2346,6 +2346,72 @@ impl GitPanel {
+                    }
+                })
+                .or_else(|| remote_origin.as_deref())
+                .or_else(|| remote_upstream.as_deref())
```
Same thing here, non-blocking but could potentially be simplified up a bit 🙂

```rust
.and_then(|upstream| match upstream.remote_name() {
    Some("upstream") => remote_upstream.as_deref(),
    Some(_) => remote_origin.as_deref(),
    None => None,
})
.or(remote_origin.as_deref())
.or(remote_upstream.as_deref())
```

dinocosta reviewed (approved):
Thank you for taking the time to tackle the feedback @tommyming ! I've just taken another look at the changes and it's looking good to me!

I've still left some comments on optional changes you could make, so let me know if you want to tackle those, otherwise I can go ahead and do that myself and get this merged 🚢

P.S.: You might want to merge `main` into this branch, as we've recently done some data migration so `cargo run` will fail in this branch in case you've got and older schema in your development database.

tommyming commented on crates/git_ui/src/git_panel.rs:
```diff
@@ -2346,6 +2346,72 @@ impl GitPanel {
+                    } else {
+                        None
+                    }
+                })
```
thanks for the heads up!
I am still getting used to Rust's chained calls, originally having thoughts about Swift's `compactMap` but I cant find something similar to that in Rust lol

tommyming commented:
> I've still left some comments on optional changes you could make, so let me know if you want to tackle those, otherwise I can go ahead and do that myself and get this merged 🚢
> 
> P.S.: You might want to merge `main` into this branch, as we've recently done some data migration so `cargo run` will fail in this branch in case you've got and older schema in your development database.

Thanks for the code enhancement review @dinocosta! 
I will make the enhancement within today, and rebase the main branch to here.

Merry Christmas!

tommyming committed:
update active repo checking, update chaining methods handling, remove
parent reference of GitHostingProviderRegistry

tommyming committed:
fix string type mismatch.

dinocosta committed:
Merge branch 'main' into feature/pull-request-dialog-disable-dismiss

dinocosta committed:
fix(git_ui): fix panic while showing error toast

Fix issue in `GitPanel.create_pull_request` that panicked when showing
an error toast. `show_error_toast` attempts to update the workspace, but
the workspace is already borrowed when handling the `CreatePullRequest`
action, causing "cannot update workspace::Workspace while it is already
being updated".

Use `cx.defer_in` to show the error toast after the current update cycle
completes.

Also remove an unused import and lowercase "Create Pull Request" to
match other commands.

dinocosta reviewed (approved):
Thank you for tackling the proposed changes @tommyming ! 

While testing these new changes, I noticed that there was a panic when `git: create pull request` failed and the error toast should be shown. I ended up pushing https://github.com/zed-industries/zed/pull/42959/commits/6873f2f3374724469c078f773783ca5bb1e260dd to handle this and removed an unused import. 

Will merge after CI/CD passes! ✌️

dinocosta committed:
test(git_hosting_providers): add tests for build_create_pull_request_url
